### ISSUE TITLE
Feign mixes up header and request param values if they are of the same name

### ISSUE THREAD

**Issue by kameshbrao** [OpenFeign/feign#702]
I have a service with the some header and request param with the same name. In such case the values of the request headers are mixed up with the values of the request parameters.

My Service is as below:

```java
@GetMapping("/get-rating-headers")
	public List<Rating> testGetRating(@RequestHeader("firstHeader") String header1,@RequestHeader("testParam1") String testHeader2,
									 @RequestParam("testParam2") String testParam2, @RequestParam("testParam1") String testParam1,
									  @RequestParam("bookId") Long bookId){}
``` 

My Feign interface looks as below:
```java
@FeignClient("http://rating-service")
interface RatingService {
	@GetMapping("/ratings/get-rating/{bookId}")
	List<Rating> findRatingsByBookId(@PathVariable("bookId") Long bookId);

	@GetMapping("/ratings/get-rating-headers")
	public List<Rating> testGetRating(@RequestHeader("firstHeader") String header1,@RequestHeader("testParam1") String testHeader2,
									  @RequestParam("testParam2") String testParam2,@RequestParam("testParam1") String testParam1,
									  @RequestParam("bookId") Long bookId);
}
```

My Client call code looks as below:
```java
@GetMapping("/{bookId}")
	public Book findBook(@RequestHeader("firstHeader") String header1,@RequestHeader("testParam1") String header2,
						 @PathVariable Long bookId) {
		System.out.println("Header1:"+header1);
		System.out.println("Header2:"+header2);
		System.out.println("BookId:"+bookId);
		List<Rating> ratings = ratingService.testGetRating(header1,header2,"testParam1","testParam2",bookId);
}
```

**Comment by kdavisk6** [comment]
I would ask this question with the `Spring` folks.  From your examples, you are using the `Spring Cloud Open Feign` annotations and not Feign core.  I would check with them first.

**Comment by kdavisk6** [comment]
I've looked into this some more and you are correct.  In `Feign` a `Param` can reference a templated value in either a `@Body`, `@Headers`, or `@RequestLine` value.  

I'm going to look into a possible long term solution for this, but to address your immediate issue, I recommend using different `Header` and `Parameter` template names.

**Comment by kdavisk6** [comment]
Closed via #778   Documentation added to call out this scenario
